File: scripts/wl138_decomposition_progress.py
#!/usr/bin/env python3
"""Emit WL-138 decomposition progress as a machine-readable artifact."""

from __future__ import annotations

import argparse
import json
import shlex
import subprocess
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

REPO_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_OUTPUT = REPO_ROOT / "docs" / "reports" / "artifacts" / "wl138_decomposition_progress.json"


CHECKPOINTS = [
    {
        "checkpoint_id": "python-monolith-cuts",
        "description": "Python command/server decomposition scaffolding exists",
        "paths": [
            "src/thegent/cli/commands/helpers.py",
            "src/thegent/mcp/server_runtime_helpers.py",
        ],
    },
    {
        "checkpoint_id": "rust-hook-splits",
        "description": "Rust hook dispatcher decomposition folders exist and execute via Rust tests",
        "paths": [
            "hooks/hook-dispatcher/src/dispatch",
            "hooks/hook-dispatcher/src/contract",
            "hooks/hook-dispatcher/src/io",
        ],
        "execution_gates": [
            {
                "gate_id": "rust-hook-dispatcher-tests",
                "description": "hook-dispatcher decomposition modules compile + tests execute",
                "command": [
                    "cargo",
                    "test",
                    "-q",
                    "--manifest-path",
                    "hooks/hook-dispatcher/Cargo.toml",
                ],
            }
        ],
    },
    {
        "checkpoint_id": "zig-abi-gate",
        "description": "Zig ABI contract + promotion checks execute",
        "paths": [
            "contracts/runtime/zig_abi_contract_v1.json",
            "scripts/validate_zig_abi_contract.py",
            "scripts/check_zig_abi_artifact.py",
            "tests/fixtures/runtime/zig_abi_symbols_fixture.txt",
            "tests/fixtures/runtime/zig_abi_error_envelope_fixture.json",
        ],
        "execution_gates": [
            {
                "gate_id": "zig-contract-validation",
                "description": "contract schema and readiness gates validate",
                "command": [
                    sys.executable,
                    "scripts/validate_zig_abi_contract.py",
                    "--contract",
                    "contracts/runtime/zig_abi_contract_v1.json",
                ],
            },
            {
                "gate_id": "zig-abi-artifact-check",
                "description": "required symbols + error envelope pass artifact check",
                "command": [
                    sys.executable,
                    "scripts/check_zig_abi_artifact.py",
                    "--contract",
                    "contracts/runtime/zig_abi_contract_v1.json",
                    "--symbols-file",
                    "tests/fixtures/runtime/zig_abi_symbols_fixture.txt",
                    "--error-envelope-json",
                    "tests/fixtures/runtime/zig_abi_error_envelope_fixture.json",
                ],
            },
        ],
    },
    {
        "checkpoint_id": "mojo-gate",
        "description": "Mojo contract + promotion gate outcome tests execute",
        "paths": [
            "contracts/runtime/mojo_kernel_contract_v1.json",
            "scripts/mojo_score_rank_harness.py",
            "tests/test_mojo_score_rank_harness.py",
        ],
        "execution_gates": [
            {
                "gate_id": "mojo-promotion-gate-outcomes",
                "description": "harness smoke + enforced promotion-gate behavior are verified",
                "command": [
                    sys.executable,
                    "-m",
                    "pytest",
                    "-q",
                    "tests/test_mojo_score_rank_harness.py::test_run_smoke_with_fake_mojo",
                    "tests/test_mojo_score_rank_harness.py::test_run_enforces_promotion_gate_by_default",
                ],
            }
        ],
    },
    {
        "checkpoint_id": "runtime-matrix-artifacts",
        "description": "Wave-2 runtime and migration artifacts are present",
        "paths": [
            "contracts/runtime/runtime-modularization-matrix.json",
            "contracts/runtime/wl131_batch_a_rust_migration_v1.json",
        ],
    },
]


def _command_result(command: list[str], root: Path) -> dict[str, Any]:
    started = time.perf_counter()
    completed = subprocess.run(command, cwd=root, capture_output=True, text=True, check=False)
    elapsed_ms = round((time.perf_counter() - started) * 1000, 2)
    return {
        "status": "pass" if completed.returncode == 0 else "fail",
        "exit_code": completed.returncode,
        "duration_ms": elapsed_ms,
        "stdout_tail": (completed.stdout or "").strip().splitlines()[-5:],
        "stderr_tail": (completed.stderr or "").strip().splitlines()[-5:],
    }


def _execution_gates_result(
    root: Path, item: dict[str, Any], *, skip_execution_gates: bool
) -> tuple[list[dict[str, Any]], bool, int, int]:
    gates = item.get("execution_gates", [])
    if not gates:
        return [], True, 0, 0

    results: list[dict[str, Any]] = []
    passed = 0
    for gate in gates:
        command = gate["command"]
        if skip_execution_gates:
            gate_result = {
                "status": "skipped",
                "exit_code": None,
                "duration_ms": 0.0,
                "stdout_tail": [],
                "stderr_tail": [],
            }
        else:
            gate_result = _command_result(command, root)
            if gate_result["status"] == "pass":
                passed += 1

        results.append(
            {
                "gate_id": gate["gate_id"],
                "description": gate["description"],
                "command": " ".join(shlex.quote(part) for part in command),
                **gate_result,
            }
        )

    total = len(gates)
    gates_complete = (passed == total) and not skip_execution_gates
    return results, gates_complete, passed, total


def _checkpoint_result(root: Path, item: dict[str, Any], *, skip_execution_gates: bool) -> dict[str, Any]:
    checks: list[dict[str, Any]] = []
    paths_complete = True
    for rel_path in item["paths"]:
        path = root / rel_path
        exists = path.exists()
        checks.append({"path": rel_path, "exists": exists})
        if not exists:
            paths_complete = False

    execution_gates, gates_complete, passed_gates, total_gates = _execution_gates_result(
        root,
        item,
        skip_execution_gates=skip_execution_gates,
    )

    complete = paths_complete and gates_complete
    return {
        "checkpoint_id": item["checkpoint_id"],
        "description": item["description"],
        "complete": complete,
        "checks": checks,
        "execution_gates": execution_gates,
        "evaluation": {
            "paths_complete": paths_complete,
            "execution_gates_complete": gates_complete,
            "passed_execution_gates": passed_gates,
            "total_execution_gates": total_gates,
            "execution_gates_skipped": skip_execution_gates,
        },
    }


def build_progress(root: Path, *, skip_execution_gates: bool) -> dict[str, Any]:
    checkpoints = [_checkpoint_result(root, item, skip_execution_gates=skip_execution_gates) for item in CHECKPOINTS]
    complete_count = sum(1 for item in checkpoints if item["complete"])
    total = len(checkpoints)
    complete_execution_gates = sum(
        checkpoint["evaluation"]["passed_execution_gates"] for checkpoint in checkpoints
    )
    total_execution_gates = sum(
        checkpoint["evaluation"]["total_execution_gates"] for checkpoint in checkpoints
    )
    return {
        "workstream_id": "WL-138",
        "artifact_id": "wl138.decomposition_progress.v1",
        "generated_at_utc": datetime.now(timezone.utc).isoformat(),
        "complete_checkpoints": complete_count,
        "total_checkpoints": total,
        "completion_pct": round((complete_count / total) * 100, 2) if total else 0.0,
        "execution_gates": {
            "complete": complete_execution_gates,
            "total": total_execution_gates,
            "completion_pct": round((complete_execution_gates / total_execution_gates) * 100, 2)
            if total_execution_gates
            else 0.0,
            "skipped": skip_execution_gates,
        },
        "checkpoints": checkpoints,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Generate WL-138 decomposition progress artifact.")
    parser.add_argument(
        "--output",
        type=Path,
        default=DEFAULT_OUTPUT,
        help="Output JSON path.",
    )
    parser.add_argument(
        "--skip-execution-gates",
        action="store_true",
        help="Skip command execution gates and emit only path-level evaluation.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    payload = build_progress(REPO_ROOT, skip_execution_gates=args.skip_execution_gates)
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(f"wrote WL-138 progress artifact: {args.output}")
    print(
        f"completion: {payload['complete_checkpoints']}/{payload['total_checkpoints']} "
        f"({payload['completion_pct']}%)"
    )
    print(
        f"execution gates: {payload['execution_gates']['complete']}/{payload['execution_gates']['total']} "
        f"({payload['execution_gates']['completion_pct']}%)"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/thegent-utils/src/bin/wl138_decomposition_progress.rs
//! Emit WL-138 decomposition progress as machine-readable JSON.

use std::{path::PathBuf, process::Command, time::Instant};

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use clap::Parser;
use serde::{Deserialize, Serialize};

#[derive(Parser)]
#[command(name = "wl138-decomposition-progress")]
struct Args {
    #[arg(
        long,
        default_value = "docs/reports/artifacts/wl138_decomposition_progress.json"
    )]
    output: PathBuf,

    #[arg(long)]
    skip_execution_gates: bool,

    #[arg(long)]
    checkpoint: Option<PathBuf>,

    #[arg(long, default_value = ".")]
    repo_root: PathBuf,

    #[arg(long, default_value = "python3")]
    python_bin: String,
}

#[derive(Clone, Serialize, Deserialize)]
struct Checkpoint {
    checkpoint_id: String,
    description: String,
    paths: Vec<String>,
    #[serde(default)]
    execution_gates: Vec<ExecutionGate>,
}

#[derive(Clone, Serialize, Deserialize)]
struct ExecutionGate {
    gate_id: String,
    description: String,
    command: Vec<String>,
}

#[derive(Serialize)]
struct CheckRecord {
    path: String,
    exists: bool,
}

#[derive(Serialize)]
struct ExecutionGateResult {
    gate_id: String,
    description: String,
    command: String,
    status: String,
    exit_code: Option<i32>,
    duration_ms: f64,
    stdout_tail: Vec<String>,
    stderr_tail: Vec<String>,
}

#[derive(Serialize)]
struct CheckpointResult {
    checkpoint_id: String,
    description: String,
    complete: bool,
    checks: Vec<CheckRecord>,
    execution_gates: Vec<ExecutionGateResult>,
    evaluation: CheckpointEvaluation,
}

#[derive(Serialize)]
struct CheckpointEvaluation {
    paths_complete: bool,
    execution_gates_complete: bool,
    passed_execution_gates: usize,
    total_execution_gates: usize,
    execution_gates_skipped: bool,
}

#[derive(Serialize)]
struct Payload {
    workstream_id: &'static str,
    artifact_id: &'static str,
    generated_at_utc: DateTime<Utc>,
    complete_checkpoints: usize,
    total_checkpoints: usize,
    completion_pct: f64,
    execution_gates: ExecutionGatesSummary,
    checkpoints: Vec<CheckpointResult>,
}

#[derive(Serialize)]
struct ExecutionGatesSummary {
    complete: usize,
    total: usize,
    completion_pct: f64,
    skipped: bool,
}

fn main() {
    if let Err(err) = run() {
        eprintln!("{err}");
        std::process::exit(1);
    }
}

fn run() -> Result<()> {
    let args = Args::parse();
    let repo_root = args.repo_root;

    if let Some(checkpoint_path) = args.checkpoint {
        let single = load_single_checkpoint(&checkpoint_path)?;
        let result = checkpoint_result(
            &repo_root,
            &single,
            args.skip_execution_gates,
            &args.python_bin,
        );
        write_json(&args.output, &result)?;
        println!("wrote WL-138 progress artifact: {}", args.output.display());
        println!(
            "completion: {}/{} ({}%)",
            usize::from(result.complete),
            1,
            if result.complete { 100.0 } else { 0.0 }
        );
        println!(
            "execution gates: {}/{} ({}%)",
            result.evaluation.passed_execution_gates,
            result.evaluation.total_execution_gates,
            percent(
                result.evaluation.passed_execution_gates,
                result.evaluation.total_execution_gates,
            )
        );
        return Ok(());
    }

    let checkpoints = build_checkpoints();
    let results = checkpoints
        .into_iter()
        .map(|item| {
            checkpoint_result(
                &repo_root,
                &item,
                args.skip_execution_gates,
                &args.python_bin,
            )
        })
        .collect::<Vec<_>>();
    let payload = Payload::for_all(results);
    write_json(&args.output, &payload)?;
    println!("wrote WL-138 progress artifact: {}", args.output.display());
    println!(
        "completion: {}/{} ({}%)",
        payload.complete_checkpoints, payload.total_checkpoints, payload.completion_pct
    );
    println!(
        "execution gates: {}/{} ({}%)",
        payload.execution_gates.complete,
        payload.execution_gates.total,
        payload.execution_gates.completion_pct
    );

    Ok(())
}

impl Payload {
    fn for_all(checkpoints: Vec<CheckpointResult>) -> Self {
        let complete_count = checkpoints.iter().filter(|cp| cp.complete).count();
        let total = checkpoints.len();
        let complete_execution_gates = checkpoints
            .iter()
            .map(|cp| cp.evaluation.passed_execution_gates)
            .sum();
        let total_execution_gates = checkpoints
            .iter()
            .map(|cp| cp.evaluation.total_execution_gates)
            .sum();
        let skipped = if total == 0 {
            false
        } else {
            checkpoints[0].evaluation.execution_gates_skipped
        };

        Payload {
            workstream_id: "WL-138",
            artifact_id: "wl138.decomposition_progress.v1",
            generated_at_utc: Utc::now(),
            complete_checkpoints: complete_count,
            total_checkpoints: total,
            completion_pct: percent(complete_count, total),
            execution_gates: ExecutionGatesSummary {
                complete: complete_execution_gates,
                total: total_execution_gates,
                completion_pct: percent(complete_execution_gates, total_execution_gates),
                skipped,
            },
            checkpoints,
        }
    }
}

fn percent(n: usize, d: usize) -> f64 {
    if d == 0 {
        0.0
    } else {
        ((n as f64) / (d as f64) * 100.0 * 100.0).round() / 100.0
    }
}

fn write_json<T: Serialize>(path: &PathBuf, value: &T) -> Result<()> {
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)
            .with_context(|| format!("failed to create output directory {}", parent.display()))?;
    }
    let content = serde_json::to_string_pretty(value)?;
    std::fs::write(path, format!("{content}\n"))
        .with_context(|| format!("failed to write output artifact to {}", path.display()))?;
    Ok(())
}

fn load_single_checkpoint(path: &PathBuf) -> Result<Checkpoint> {
    let raw = std::fs::read_to_string(path)
        .with_context(|| format!("failed to read checkpoint fixture {}", path.display()))?;
    let checkpoint = serde_json::from_str::<Checkpoint>(&raw)
        .with_context(|| format!("failed to parse checkpoint JSON {}", path.display()))?;
    Ok(checkpoint)
}

fn checkpoint_result(
    root: &PathBuf,
    item: &Checkpoint,
    skip_execution_gates: bool,
    python_bin: &str,
) -> CheckpointResult {
    let mut checks = Vec::with_capacity(item.paths.len());
    let mut paths_complete = true;
    for rel_path in &item.paths {
        let path = root.join(rel_path);
        let exists = path.exists();
        if !exists {
            paths_complete = false;
        }
        checks.push(CheckRecord {
            path: rel_path.clone(),
            exists,
        });
    }

    let mut gate_results = Vec::with_capacity(item.execution_gates.len());
    let mut passed = 0usize;

    for gate in item.execution_gates.iter().cloned().collect::<Vec<_>>() {
        let command = normalize_command(gate.clone(), python_bin);
        let result = if skip_execution_gates {
            ExecutionGateResult {
                gate_id: gate.gate_id,
                description: gate.description,
                command: command.join(" "),
                status: "skipped".to_string(),
                exit_code: None,
                duration_ms: 0.0,
                stdout_tail: Vec::new(),
                stderr_tail: Vec::new(),
            }
        } else {
            let (status, exit_code, duration_ms, stdout_tail, stderr_tail) =
                execute_command(root, &command);
            if status == "pass" {
                passed += 1;
            }
            ExecutionGateResult {
                gate_id: gate.gate_id,
                description: gate.description,
                command: command.join(" "),
                status,
                exit_code,
                duration_ms,
                stdout_tail,
                stderr_tail,
            }
        };
        gate_results.push(result);
    }

    let total = gate_results.len();
    let gates_complete = if total == 0 {
        true
    } else {
        passed == total && !skip_execution_gates
    };

    CheckpointResult {
        checkpoint_id: item.checkpoint_id.clone(),
        description: item.description.clone(),
        complete: paths_complete && gates_complete,
        checks,
        execution_gates: gate_results,
        evaluation: CheckpointEvaluation {
            paths_complete,
            execution_gates_complete: gates_complete,
            passed_execution_gates: passed,
            total_execution_gates: total,
            execution_gates_skipped: skip_execution_gates,
        },
    }
}

fn normalize_command(gate: ExecutionGate, python_bin: &str) -> Vec<String> {
    let mut command = gate.command;
    if command.first().is_some_and(|first| first == "${PYTHON}") {
        command[0] = python_bin.to_string();
    }
    command
}

fn execute_command(
    root: &PathBuf,
    command: &[String],
) -> (String, Option<i32>, f64, Vec<String>, Vec<String>) {
    if command.is_empty() {
        return (
            "fail".to_string(),
            Some(127),
            0.0,
            Vec::new(),
            vec!["empty command".to_string()],
        );
    }

    let start = Instant::now();
    let output = Command::new(&command[0])
        .args(&command[1..])
        .current_dir(root)
        .output();
    let elapsed_ms = (Instant::now() - start).as_secs_f64() * 1000.0;

    match output {
        Ok(completed) => {
            let status = if completed.status.success() {
                "pass"
            } else {
                "fail"
            }
            .to_string();

            let exit_code = completed.status.code();
            let stdout_tail = tail_lines(String::from_utf8_lossy(&completed.stdout).to_string());
            let stderr_tail = tail_lines(String::from_utf8_lossy(&completed.stderr).to_string());
            (status, exit_code, elapsed_ms, stdout_tail, stderr_tail)
        }
        Err(err) => (
            "fail".to_string(),
            Some(1),
            elapsed_ms,
            Vec::new(),
            vec![err.to_string()],
        ),
    }
}

fn tail_lines(data: String) -> Vec<String> {
    let mut lines = data.lines().map(ToOwned::to_owned).collect::<Vec<_>>();
    if lines.len() > 5 {
        lines.drain(0..(lines.len() - 5));
    }
    lines
}

fn build_checkpoints() -> Vec<Checkpoint> {
    vec![
        Checkpoint {
            checkpoint_id: "python-monolith-cuts".to_string(),
            description: "Python command/server decomposition scaffolding exists".to_string(),
            paths: vec![
                "src/thegent/cli/commands/helpers.py".to_string(),
                "src/thegent/mcp/server_runtime_helpers.py".to_string(),
            ],
            execution_gates: Vec::new(),
        },
        Checkpoint {
            checkpoint_id: "rust-hook-splits".to_string(),
            description: "Rust hook dispatcher decomposition folders exist and execute via Rust tests".to_string(),
            paths: vec![
                "hooks/hook-dispatcher/src/dispatch".to_string(),
                "hooks/hook-dispatcher/src/contract".to_string(),
                "hooks/hook-dispatcher/src/io".to_string(),
            ],
            execution_gates: vec![
                ExecutionGate {
                    gate_id: "rust-hook-dispatcher-tests".to_string(),
                    description: "hook-dispatcher decomposition modules compile + tests execute".to_string(),
                    command: vec![
                        "cargo".to_string(),
                        "test".to_string(),
                        "-q".to_string(),
                        "--manifest-path".to_string(),
                        "hooks/hook-dispatcher/Cargo.toml".to_string(),
                    ],
                },
            ],
        },
        Checkpoint {
            checkpoint_id: "zig-abi-gate".to_string(),
            description: "Zig ABI contract + promotion checks execute".to_string(),
            paths: vec![
                "contracts/runtime/zig_abi_contract_v1.json".to_string(),
                "scripts/validate_zig_abi_contract.py".to_string(),
                "scripts/check_zig_abi_artifact.py".to_string(),
                "tests/fixtures/runtime/zig_abi_symbols_fixture.txt".to_string(),
                "tests/fixtures/runtime/zig_abi_error_envelope_fixture.json".to_string(),
            ],
            execution_gates: vec![
                ExecutionGate {
                    gate_id: "zig-contract-validation".to_string(),
                    description: "contract schema and readiness gates validate".to_string(),
                    command: vec![
                        "${PYTHON}".to_string(),
                        "scripts/validate_zig_abi_contract.py".to_string(),
                        "--contract".to_string(),
                        "contracts/runtime/zig_abi_contract_v1.json".to_string(),
                    ],
                },
                ExecutionGate {
                    gate_id: "zig-abi-artifact-check".to_string(),
                    description: "required symbols + error envelope pass artifact check".to_string(),
                    command: vec![
                        "${PYTHON}".to_string(),
                        "scripts/check_zig_abi_artifact.py".to_string(),
                        "--contract".to_string(),
                        "contracts/runtime/zig_abi_contract_v1.json".to_string(),
                        "--symbols-file".to_string(),
                        "tests/fixtures/runtime/zig_abi_symbols_fixture.txt".to_string(),
                        "--error-envelope-json".to_string(),
                        "tests/fixtures/runtime/zig_abi_error_envelope_fixture.json".to_string(),
                    ],
                },
            ],
        },
        Checkpoint {
            checkpoint_id: "mojo-gate".to_string(),
            description: "Mojo contract + promotion gate outcome tests execute".to_string(),
            paths: vec![
                "contracts/runtime/mojo_kernel_contract_v1.json".to_string(),
                "scripts/mojo_score_rank_harness.py".to_string(),
                "tests/test_mojo_score_rank_harness.py".to_string(),
            ],
            execution_gates: vec![
                ExecutionGate {
                    gate_id: "mojo-promotion-gate-outcomes".to_string(),
                    description:
                        "harness smoke + enforced promotion-gate behavior are verified".to_string(),
                    command: vec![
                        "${PYTHON}".to_string(),
                        "-m".to_string(),
                        "pytest".to_string(),
                        "-q".to_string(),
                        "tests/test_mojo_score_rank_harness.py::test_run_smoke_with_fake_mojo".to_string(),
                        "tests/test_mojo_score_rank_harness.py::test_run_enforces_promotion_gate_by_default".to_string(),
                    ],
                },
            ],
        },
        Checkpoint {
            checkpoint_id: "runtime-matrix-artifacts".to_string(),
            description: "Wave-2 runtime and migration artifacts are present".to_string(),
            paths: vec![
                "contracts/runtime/runtime-modularization-matrix.json".to_string(),
                "contracts/runtime/wl131_batch_a_rust_migration_v1.json".to_string(),
            ],
            execution_gates: Vec::new(),
        },
    ]
}
